### PR TITLE
[SPARK-4086][GraphX]: Fold-style aggregation for VertexRDD

### PR DESCRIPTION
**Pull request by brennonyork**
Adds five new methods into the `VertexRDD` suite to allow for fold-style calling conventions. Those methods are:
- `leftZipJoinWithFold`
- `leftJoinWithFold`
- `innerZipJoinWithFold`
- `innerJoinWithFold`
- `aggregateUsingIndexWithFold`

Each of above has a set of tests within the `VertexRDDSuite` to ensure proper functionality.

### CODE REVIEW

**Comment by brennonyork** [comment]
@ankurdave thoughts / comments? Is this in line with what you were looking for?

File: graphx/src/main/scala/org/apache/spark/graphx/VertexRDD.scala [hunk 1]
```diff
@@ -154,7 +154,30 @@ abstract class VertexRDD[VD](
...
    */
   def leftZipJoin[VD2: ClassTag, VD3: ClassTag]
-      (other: VertexRDD[VD2])(f: (VertexId, VD, Option[VD2]) => VD3): VertexRDD[VD3]
+      (other: VertexRDD[VD2])
+      (f: (VertexId, VD, Option[VD2]) => VD3)
+    : VertexRDD[VD3]
+
+  /**
+   * Left joins this RDD with another VertexRDD with the same index. This function will fail if
+   * both VertexRDDs do not share the same index. The resulting vertex set contains an entry for
```
Comment (maropu) [inline graphx/src/main/scala/org/apache/spark/graphx/VertexRDD.scala]: Is "will fail" correct? In the implementation, I think the codes fail back to creating same indices.

File: graphx/src/main/scala/org/apache/spark/graphx/VertexRDD.scala [hunk 2]
```diff
@@ -154,7 +154,30 @@ abstract class VertexRDD[VD](
...
+   *
+   * @param other the other VertexRDD with which to join.
+   * @param acc the initial value for the accumulator
+   * @param f the function mapping a vertex id and its attributes in this and the other vertex set
+   * to a new vertex attribute.
+   * @return a VertexRDD containing the results of `f`
+   */
+  def leftZipJoinWithFold[VD2: ClassTag, VD3: ClassTag, A]
+      (other: VertexRDD[VD2], acc: A)
+      (f: (A, VertexId, VD, Option[VD2]) => VD3)
```
Comment (maropu) [inline graphx/src/main/scala/org/apache/spark/graphx/VertexRDD.scala]: ISTM, not 'acc' but 'initVal' would  be better for an initial value.

File: graphx/src/main/scala/org/apache/spark/graphx/impl/VertexPartitionBaseOps.scala [hunk 3]
```diff
@@ -136,6 +136,31 @@ private[graphx] abstract class VertexPartitionBaseOps
     leftJoin(createUsingIndex(other))(f)
   }
 
+  def leftJoinWithFold[VD2: ClassTag, VD3: ClassTag, A]
+      (other: Self[VD2], acc: A)
```
Comment (maropu) [inline graphx/src/main/scala/org/apache/spark/graphx/impl/VertexPartitionBaseOps.scala]: Reduce-style aggregation is named 'leftJoin', and fold-style aggregation is 'leftJoinWithFold'.
A rule of naming is kind of inconsistent, so would it be better to rename 'leftJoin' into 'leftJoinWithReduce'? Thought?

Comment (brennonyork) [inline graphx/src/main/scala/org/apache/spark/graphx/impl/VertexPartitionBaseOps.scala]: I'm pretty sure @ankurdave was looking at the `*WithFold` style addition and not changing the original method names to keep backwards compatibility as best as possible. I agree its a bit confusing, but I figure maintaining backwards compatibility as best as best is the better option.

File: graphx/src/main/scala/org/apache/spark/graphx/VertexRDD.scala (hunk 2 above)
Comment (brennonyork) [inline graphx/src/main/scala/org/apache/spark/graphx/VertexRDD.scala]: Very good point. Will update that.

File: graphx/src/main/scala/org/apache/spark/graphx/VertexRDD.scala (hunk 1 above)
Comment (brennonyork) [inline graphx/src/main/scala/org/apache/spark/graphx/VertexRDD.scala]: Very true. Rereading the docs it looks like they haven't been updated with verbage of a few of the more recent bug fixes. Will add that.

**Comment by maropu** [comment]
lgtm except for the naming issue.

**Comment by brennonyork** [comment]
jenkins, retest this please

**Comment by ankurdave** [comment]
Hmm, this doesn't really fulfill the role of folds at the moment. The initial values are passed to a "fold" function that otherwise does the same thing as a reduce function, so it's equivalent to the user just generating the initial values in the fold function themselves.

With SPARK-4086, I think @jegonzal and I were imagining a use case where you have an Iterator[(VertexId, A)] with duplicate keys and you want a VertexRDD[Array[A]]. Currently the only way is to map each element to an array and then concatenate arrays, which creates a lot of garbage. With folds you could use an empty ArrayBuffer as the initial value and do in-place inserts as elements appear.

So it only seems to make sense to have a fold version of `aggregateUsingIndex` and the `VertexRDD.apply` constructor (maybe called `VertexRDD.createWithFold`), not the joins, because joining two VertexRDDs will never involve more than two values for a key.

Also, in general the fold function has to return a value of the same type as the init value, e.g. `def fold[A](initVal: A, f: (VD, A) => A): VertexRDD[A]`.

**Comment by brennonyork** [comment]
@ankurdave, thanks for the clarification. Let me take a second stab at this given what you stated and I should have something much more in line with the original thought!

**Comment by brennonyork** [comment]
jenkins retest this please

**Comment by brennonyork** [comment]
@ankurdave when you get a chance can you review this? I know its been a while, but I finally had a chance to get back to this and rework it given your above comments. The only issue I had was with the Iterator[(VertexId, A)] and, instead, assumed a structure of RDD[(VertexId, A)] with, possibly, duplicate keys.

**Comment by brennonyork** [comment]
jenkins retest this please

**Comment by brennonyork** [comment]
jenkins retest this please

**Comment by andrewor14** [comment]
@ankurdave any thoughts on this?

**Comment by srowen** [comment]
(Is GraphX active anymore? I haven't seen activity or responses on these issues in quite a while)

**Comment by jegonzal** [comment]
@srowen GraphX is still active we have just been pretty busy with some other changes.  Let me see what needs to be done with this PR.

**Comment by rxin** [comment]
I'm going to close this pull request. If this is still relevant and you are interested in pushing it forward, please open a new pull request. Thanks!

**Pull request closed without merging**